### PR TITLE
Add Python factorial Interpreter

### PR DESCRIPTION
**Pull request by Salvatore112**

### CODE REVIEW

**Comment by Kakadu** [comment]
Linter report from 2023-12-07 17:58, for mini language Python
```
File '_build/default/lib/tests.ml' doesn't have corresponding .mli interface

File "lib/tests.ml", line 1, characters 0-8:
1 | open Ast
    ^^^^^^^^
Alert zanuda-linter:  First item in file should be a documentation comment with copyright information. For example:
            (** Copyright 2021-2022, Winnie Pooh et al. *)
File "lib/tests.ml", line 2, characters 0-11:
2 | open Parser
    ^^^^^^^^^^^
Alert zanuda-linter: Second item in file should be a documentation comment with correct license information. For example:
           (** SPDX-License-Identifier: LGPL-3.0-or-later *)
File "lib/tests.ml", line 8, characters 9-17:
8 |   | _ -> failwith "Failed to unpack"
             ^^^^^^^^
Alert zanuda-linter: Using `failwith` (or `assert false`) usually is a clue that a corner case is not being handled properly. To report errors we recommend using error monad instead. In princliple, these construction are OK for temporary work-in-progress code, but in release they should be eliminated
File "lib/interpreter.ml", lines 268-274, characters 34-32:
268 | ..................................expression_list =
269 |              match expression_list with
270 |              | [] -> return None
271 |              | exp :: tl ->
272 |                let* value = i_expr exp_or_stmt env exp in
273 |                let () = print_string (pack_to_string value) in
274 |                print_exp_list tl
Alert zanuda-linter: Using `function` is recommended
File "lib/interpreter.ml", lines 268-274, characters 34-32:
268 | ..................................expression_list =
269 |              match expression_list with
270 |              | [] -> return None
271 |              | exp :: tl ->
272 |                let* value = i_expr exp_or_stmt env exp in
273 |                let () = print_string (pack_to_string value) in
274 |                print_exp_list tl
Alert zanuda-linter: Using `function` is recommended
File "lib/interpreter.mli", lines 1-7, characters 0-3:
1 | module type MONAD = sig
2 |   type 'a t
3 | 
4 |   val return : 'a -> 'a t
5 |   val ( >>= ) : 'a t -> ('a -> 'b t) -> 'b t
6 |   val ( let* ) : 'a t -> ('a -> 'b t) -> 'b t
7 | end
Alert zanuda-linter:  First item in file should be a documentation comment with copyright information. For example:
            (** Copyright 2021-2022, Winnie Pooh et al. *)
File "lib/interpreter.mli", lines 9-16, characters 0-3:
 9 | module type MONADERROR = sig
10 |   type 'a t
11 | 
12 |   val return : 'a -> 'a t
13 |   val ( >>= ) : 'a t -> ('a -> 'b t) -> 'b t
14 |   val ( let* ) : 'a t -> ('a -> 'b t) -> 'b t
15 |   val error : string -> 'a t
16 | end
Alert zanuda-linter: Second item in file should be a documentation comment with correct license information. For example:
           (** SPDX-License-Identifier: LGPL-3.0-or-later *)
```

File: Python/lib/interpreter.ml [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+  let get_func i env =
+    List.find (fun (x : function_symb) -> x.identifier = i) env.functions
+  ;;
+
+  (* Miscellaneous *)
+  let pack_to_string = function
+    | String a -> a
+    | Int a -> Int.to_string a
+    | Bool a -> Bool.to_string a
+    | _ -> "Interpretation Error"
```
Comment (Kakadu) [inline Python/lib/interpreter.ml:148]: cringe

File: Python/lib/interpreter.ml [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                { (change_or_add_var_list
+                     env
+                     (combine_args_and_params (get_func identifier env).params x))
+                  with
+                  id = Identifier "local"
+                ; local_envs = [ { local_env with vars = env.vars } ]
+                ; functions = [ get_func identifier env ]
+                }
+                (get_func identifier env).body))
+      | _ -> error "wip"
```
Comment (Kakadu) [inline Python/lib/interpreter.ml:293]: ?

File: Python/lib/tests.ml [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                           ) ))
+                ] )
+          ] )
+    ; Expression
+        (FunctionCall
+           ( Identifier "print"
+           , [ FunctionCall (Identifier "factorial", [ Const (Int 5) ]) ] ))
+    ]
+;;
+
```
Comment (Kakadu) [inline Python/lib/tests.ml:227]: Сделайте уже cram  testы, в этом невозможно ничего понять.

File: Python/lib/interpreter.ml [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        let* a = i_expr exp_or_stmt env a in
+        let* b = i_expr exp_or_stmt env b in
+        return (Bool (a = Bool true || b = Bool true))
+      | Variable i ->
+        (match var_in_env i env with
+         | true -> return (get_var i env).value
+         | false -> error "undefined Variable")
+      | FunctionCall (identifier, exp_list) ->
+        (match identifier with
+         | Identifier "print" ->
```
Comment (Kakadu) [inline Python/lib/interpreter.ml:267]: Что будет, Если человек переопределит print?

File: Python/lib/interpreter.ml (hunk 1 above)
Comment (Salvatore112) [reply Python/lib/interpreter.ml:148]: Почему?

File: Python/lib/interpreter.ml (hunk 4 above)
Comment (Salvatore112) [reply Python/lib/interpreter.ml:267]: Выглядит необычно, но на переопределение print интерпретатор нормально реагирует.
image

**Comment by Kakadu** [comment]
Linter report from 2023-12-08 21:00, for mini language Python
```
```

File: Python/lib/interpreter.ml (hunk 1 above)
Comment (Kakadu) [reply Python/lib/interpreter.ml:148]: Вы проглатываете ошибку. Тут по идее монада должна по существу использоваться.

**Comment by Kakadu** [comment]
Под cram тестами я имел в виду, что пишется текст на питоне, а интерпретатор его интерпретирует. А у вас парсер никак не используется.

File: Python/lib/interpreter.ml (hunk 4 above)
Comment (Kakadu) [reply Python/lib/interpreter.ml:267]: Я хочу тест на это. Можно в следующем PR.

File: Python/lib/interpreter.ml (hunk 2 above)
Comment (Kakadu) [reply Python/lib/interpreter.ml:293]: Ну, потом исправите.

**Comment by Kakadu** [comment]
Я думаю, недоработки можно отложить до следующего PR. OK?

**Comment by Salvatore112** [comment]
> Я думаю, недоработки можно отложить до следующего PR. OK?

В смысле одобрить этот, а исправить в следующем? Я непротив

**Comment by Kakadu** [comment]
> > Я думаю, недоработки можно отложить до следующего PR. OK?
> 
> В смысле одобрить этот, а исправить в следующем? Я непротив

Если вы не забудете про замечания в этом, я тоже не против.
